src/number.rs

```diff
@@ -279,6 +279,17 @@ impl Number {
         }
     }
 
+    #[cfg(feature = "arbitrary_precision")]
+    /// Returns the `&str` representation of the `Number`.
+    /// ```
+    /// # use serde_json::Number;
+    ///
+    /// assert_eq!(Number::from_f64(256.0).unwrap().as_str(), "256.0");
+    /// assert_eq!(Number::from_f64(34.0).unwrap().as_str(), "34.0");
```

Comment:
Ah, that makes total sense sense. Thank you for the wisdom; I'm horrible at documentation and I'm definitely trying to get better at it. The rust ecosystem has certainly raised the bar on such.

I hope the example I just committed is more in line with what you are thinking. Please let me know if there's more I can do to improve it.

Also, this isn't relevant to this pull request but `"012.34"` is being rejected as an invalid number:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid number", line: 1, column: 2)', src/number.rs:15:54
stack backtrace:
   0:        0x104bfca34 - std::backtrace_rs::backtrace::libunwind::trace::h35c35bd6a1eea1a3
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:        0x104bfca34 - std::backtrace_rs::backtrace::trace_unsynchronized::h6e6a37fdc09ec2da
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x104bfca34 - std::sys_common::backtrace::_print_fmt::h6ae2889d87a644d1
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:65:5
   3:        0x104bfca34 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hb75be2413dd8ceb3
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:44:22
   4:        0x104c12100 - core::fmt::rt::Argument::fmt::hfc9103857ff63de2
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/fmt/rt.rs:138:9
   5:        0x104c12100 - core::fmt::write::h95d50546e769656f
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/fmt/mod.rs:1094:21
   6:        0x104bfad4c - std::io::Write::write_fmt::hce019ca594763835
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/io/mod.rs:1713:15
   7:        0x104bfc888 - std::sys_common::backtrace::_print::h15dce0f07dfee3db
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:47:5
   8:        0x104bfc888 - std::sys_common::backtrace::print::hae7307dcada41b2a
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:34:9
   9:        0x104bfdc04 - std::panicking::default_hook::{{closure}}::h30c41986d637ef23
  10:        0x104bfda0c - std::panicking::default_hook::h81d03189ef2e7e78
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:288:9
  11:        0x104bfe074 - std::panicking::rust_panic_with_hook::he8360f4d28da55fc
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:705:13
  12:        0x104bfdf80 - std::panicking::begin_panic_handler::{{closure}}::he9e07d605072520f
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:597:13
  13:        0x104bfce14 - std::sys_common::backtrace::__rust_end_short_backtrace::he03bf52f9c1eb73a
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:151:18
  14:        0x104bfdcf4 - rust_begin_unwind
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:593:5
  15:        0x104c1a3cc - core::panicking::panic_fmt::ha5f5ca2311fff2ad
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:67:14
  16:        0x104c1a744 - core::result::unwrap_failed::h34067e5e60f3683e
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/result.rs:1651:5
  17:        0x104bd0a74 - core::result::Result<T,E>::unwrap::hacd62d4f71c8f44b
  18:        0x104bd2524 - rust_out::main::_doctest_main_src_number_rs_284_0::h7692f3eb47846b9e
  19:        0x104bd234c - rust_out::main::h9d98ab4a1af73092
  20:        0x104bd03c0 - core::ops::function::FnOnce::call_once::h3582cd9180f91c06
  21:        0x104bd00b0 - std::sys_common::backtrace::__rust_begin_short_backtrace::h322056fafa780450
  22:        0x104bd011c - std::rt::lang_start::{{closure}}::hc7fab609c73091e5
  23:        0x104bf8708 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h0689b9cc840db667
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/ops/function.rs:284:13
  24:        0x104bf8708 - std::panicking::try::do_call::h8d21b0c0c04af112
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:500:40
  25:        0x104bf8708 - std::panicking::try::h618481d45c1b815c
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:464:19
  26:        0x104bf8708 - std::panic::catch_unwind::hbdeff70f3984ee7b
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panic.rs:142:14
  27:        0x104bf8708 - std::rt::lang_start_internal::{{closure}}::haa4994ba13a3cd15
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/rt.rs:148:48
  28:        0x104bf8708 - std::panicking::try::do_call::h39b55541875d339a
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:500:40
  29:        0x104bf8708 - std::panicking::try::h93ac0a218f84acad
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:464:19
  30:        0x104bf8708 - std::panic::catch_unwind::h07a4f62359dfd8f0
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panic.rs:142:14
  31:        0x104bf8708 - std::rt::lang_start_internal::hdd06e3566639fc5b
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/rt.rs:148:20
  32:        0x104bd00f4 - std::rt::lang_start::h8e60286336dfb70e
  33:        0x104bd2608 - _main
```

Reply:
The updated example is:

```rust
use serde_json::Number;
for value in [
    "7",
    "12.34",
    "34e-56789",
    "0.0123456789000000012345678900000001234567890000123456789",
    "343412345678910111213141516171819202122232425262728293034",
    "-343412345678910111213141516171819202122232425262728293031",
] {
    let number: Number = serde_json::from_str(value).unwrap();
    assert_eq!(number.as_str(), value);
}
```

Reply:
One last thing, it'd be great if `Value` had an `as_number` method. If you're interested, I can add it to this pull request or I can create a separate one.

Reply:
> This is pretty different from the intended usage of as_str(), right? I don't see why one would use as_str in this manner. From #1066 I'd inferred it would be to inspect the original formatting of a number that was parsed from a JSON document using serde_json::from_str.

The reason I personally need `as_str` is to avoid an allocation  via `to_string`. 

For a bit of context, I'm working on a JSON Schema crate. In order to properly evaluate decimals and also to support big numeric types, I have written parsers to convert strings to `num::BigRational` and `num::BigInt`.

I hope the example illustrates the utility of having access to the underlying string. If others need to compare fractions or deal with big numbers, `"arbitrary_precision"` with a custom string parser is, so far as I can tell, the only viable path to achieving said goal.